### PR TITLE
Add a authentication method based on password and make usernames unique

### PR DESCRIPTION
**Pull request by frcroth**

### CODE REVIEW

**Comment by frcroth** [comment]
@sirkrypt0 I wanted to test password verification and wrote a one line function for it. Then clippy was angry because the function wasn't used. So I implemented some basic authentication.

**Comment by frcroth** [comment]
Also this authentication method requires the user to know the id of the account which is weird. Could be saved by making username unique. That way you could just post a username password pair to users/auth and authenticate.

**Comment by sirkrypt0** [comment]
> Also this authentication method requires the user to know the id of the account which is weird. Could be saved by making username unique. That way you could just post a username password pair to users/auth and authenticate.

I like making the username unique.

**Comment by frcroth** [comment]
I have not marked this as ready, because the auth_token is not implemented, but I think a review would be beneficial, as a lot of stuff has changed @sirkrypt0

**Comment by sirkrypt0** [comment]
> I have not marked this as ready, because the auth_token is not implemented, but I think a review would be beneficial, as a lot of stuff has changed @sirkrypt0 

I will do so. We can also merge this already and implement auth tokens in another mr.

File: server/src/user/auth.rs
Comment (sirkrypt0) [inline server/src/user/auth.rs:27]: Can we use the `?` here to propagate the error?

File: server/src/user/model.rs
Comment (sirkrypt0) [inline server/src/user/model.rs:29]: Is meant to reflect all errors occurring in the user model?

File: server/src/user/model.rs
Comment (sirkrypt0) [inline server/src/user/model.rs]: What happens if `User::find_by_username` fails with e.g. a database error? Will this be propagated correctly?

According to [this](https://stackoverflow.com/questions/36362020/what-is-unwrap-in-rust-and-what-is-it-used-for) StackOverflow question it would panic which we probably don't want here.

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]

File: server/src/user/model.rs
Comment (frcroth) [reply server/src/user/model.rs:29]: Yes

Comment (frcroth) [reply server/src/user/model.rs:29]: To give some further information when something goes wrong in the Model and also be able to react to some errors that are not bad (UsernameTaken)

File: server/src/user/model.rs
Comment (frcroth) [reply server/src/user/model.rs]: Yes currently it will panic, I can change the return value to a result

**Comment by frcroth** [comment]
Ok will merge after implementing your suggestions

File: server/src/user/auth.rs
Comment (frcroth) [reply server/src/user/auth.rs:27]: This would require implementing a From for UserError -> AuthenticationError. I think the current way is better.

File: server/src/user/model.rs
Comment (frcroth) [reply server/src/user/model.rs]: Done

**Review by sirkrypt0 — APPROVED** [review]
I'd appreciate if you could wait merging this until #16 is merged and add some tests afterwards.

**Review by sirkrypt0 — APPROVED** [review]

**Pull request merged**
